src/WebJobs.Script.WebHost/Diagnostics/ConsoleWriter.cs

```diff
@@ -0,0 +1,169 @@
+                }
+                else
+                {
+                    _writeEvent = WriteToConsoleBuffer;
+                    _consoleBufferTimeout = consoleBufferTimeout;
+                    if (autoStart)
+                    {
+                        _consoleBufferReadLoop = environment.GetEnvironmentVariable(EnvironmentSettingNames.ConsoleLoggingBufferBatched) switch
+                        {
+                            "0" => ProcessConsoleBufferAsync(),             // disable batching by setting to 0
+                            _ => ProcessConsoleBufferBatchedAsync(),        // default behavior is batched
+                        };
+                    }
+                }
+            }
+            else
+            {
+                _writeEvent = (string s) => { };
+            }
+
+            _exceptionhandler = exceptionHandler;
+        }
+
+        public Action<string> WriteHandler => _writeEvent;
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            try
+            {
+                while (_consoleBuffer.Writer.TryWrite(evt) == false)
+                {
+                    // Buffer is currently full, wait until writing is permitted.
+                    // This is the downside of using channels, we are on a sync code path and so we have to block on this task
+                    var writeTask = _consoleBuffer.Writer.WaitToWriteAsync().AsTask();
+                    if (writeTask.WaitAsync(_consoleBufferTimeout).Result == false)
+                    {
+                        // The buffer has been completed and does not allow further writes.
+                        // Currently this should not be possible, but the safest thing to do is just write directly to the console.
+                        Console.WriteLine(evt);
+                        break;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                // Most likely a task cancellation exception from the timeout expiring, but regardless we use the handler
+                // dump the raw exception and write the event to the console directly
+                _exceptionhandler(ex);
+                Console.WriteLine(evt);
+            }
+        }
+
+        internal async Task ProcessConsoleBufferAsync()
+        {
+            try
+            {
+                await foreach (var line in _consoleBuffer.Reader.ReadAllAsync())
+                {
+                    Console.WriteLine(line);
+                }
+            }
+            catch (Exception ex)
+            {
+                // Not sure what can fail here, but there is nothing else monitoring this task so just log the exception
+                _exceptionhandler(ex);
+            }
+            finally
+            {
+                // if this has failed for any reason, complete the buffer so that everything falls back to console
+                _consoleBuffer.Writer.Complete();
```

Comment:
I agree on these fallbacks ultimately using Console.Writeline...but if I understand correctly it'll be via a try/fail/retry pattern on getting there. We probably want to replace the delegate with a direct Console write in this failure case (common method for both?) so that the fallback isn't a net worse over current.

Reply:
Makes sense. I updated the implementation to have an extra level of indirection so that the delegate can be replaced on the fly without the consumer (LinuxContainerEventGenerator) knowing, and the finally block now does the replacement.

Reply:
Makes sense, done.